File: leetcode-practice-java/src/test/java/io/hank/leetcode/practices/_005_Longest_Palindromic_Substring_Test.java
package io.hank.leetcode.practices;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class _005_Longest_Palindromic_Substring_Test {
    @Test
    void testLongestPalindrome_NormalCase() {
        _005_Longest_Palindromic_Substring solution = new _005_Longest_Palindromic_Substring();
        String input = "babad";
        String result = solution.longestPalindrome(input);
        assertTrue(result.equals("bab") || result.equals("aba"));
    }

    @Test
    void testLongestPalindrome_NoPalindromeSubstring() {
        _005_Longest_Palindromic_Substring solution = new _005_Longest_Palindromic_Substring();
        String input = "abcde";
        String result = solution.longestPalindrome(input);
        // Any single character is a valid palindrome
        assertTrue(result.equals("a") || result.equals("b") || result.equals("c") || result.equals("d") || result.equals("e"));
    }

    @ParameterizedTest
    @ValueSource(strings = {"a", "aaaa", "bb", ""})
    void testLongestPalindrome_EqualsSelf(String arg) {
        _005_Longest_Palindromic_Substring solution = new _005_Longest_Palindromic_Substring();
        String result = solution.longestPalindrome(arg);
        assertEquals(arg, result);
    }

    @Test
    void testLongestPalindrome_LongString() {
        _005_Longest_Palindromic_Substring solution = new _005_Longest_Palindromic_Substring();
        String input = "abcbaabcd";
        String result = solution.longestPalindrome(input);
        assertEquals("cbaabc", result);
    }
}

File: leetcode-practice-kotlin/src/test/kotlin/io/hank/leetcode/kotlin/practices/_005_Longest_Palindromic_Substring_Test.kt
package io.hank.leetcode.kotlin.practices

import kotlin.test.Test
import kotlin.test.assertEquals

class _005_Longest_Palindromic_Substring_Test {

    private val solution = _005_Longest_Palindromic_Substring()

    @Test
    fun `test with basic palindrome`() {
        val result = solution.longestPalindrome("babad")
        // The longest palindrome could be either "bab" or "aba"
        assertEquals(true, result == "bab" || result == "aba")
    }

    @Test
    fun `test with single character string`() {
        val result = solution.longestPalindrome("a")
        assertEquals("a", result)
    }

    @Test
    fun `test with entire string as palindrome`() {
        val result = solution.longestPalindrome("racecar")
        assertEquals("racecar", result)
    }

    @Test
    fun `test with string of length 2`() {
        val result = solution.longestPalindrome("aa")
        assertEquals("aa", result)
    }

    @Test
    fun `test with no palindrome longer than 1`() {
        val result = solution.longestPalindrome("abc")
        // Each character is a palindrome by itself
        assertEquals(true, result in listOf("a", "b", "c"))
    }

    @Test
    fun `test with even length palindrome`() {
        val result = solution.longestPalindrome("cbbd")
        assertEquals("bb", result)
    }

    @Test
    fun `test with empty string`() {
        val result = solution.longestPalindrome("")
        assertEquals("", result)
    }
}
